Checkpoint httpapi plugin

rcarrillocruz opened the pull request:
##### SUMMARY

Add Checkpoint httpapi plugin and module_utils

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
plugins/httpapi/checkpoint

rcarrillocruz committed:
Add checkpoint httpapi plugin and access rule facts module

rcarrillocruz committed:
WIP checkpoint_access_rule module

rcarrillocruz committed:
Add publish and install policy, plus fix empty json object request for publish

rcarrillocruz committed:
Refactor publish and install_policy onto module_utils

rcarrillocruz committed:
Add update resource logic

rcarrillocruz committed:
Add checkpoint_host_facts module

rcarrillocruz committed:
Return code and response on get_acess_rule function

rcarrillocruz committed:
Add checkpoint_host module

rcarrillocruz committed:
Add checkpoint_run_script module

rcarrillocruz committed:
Add checkpoint_task_facts module

rcarrillocruz committed:
Show all tasks if no task id is passed

Note, this is only available on v1.3 of Checkpoint WS API

rcarrillocruz committed:
Add update logic to checkpoint host

rcarrillocruz committed:
Add full details on get task call

rcarrillocruz committed:
Add checkpoint httpapi plugin

rcarrillocruz committed:
Fix pep8

Qalthos commented on lib/ansible/plugins/httpapi/checkpoint.py:
```diff
@@ -0,0 +1,137 @@
+        if username and password:
+            payload = {'user': username, 'password': password}
+            url = '/web_api/login'
+            response, response_data = self._send_auth_request(
```
You can _probably_ get away with a normal `send_request` here

Qalthos commented on lib/ansible/plugins/httpapi/checkpoint.py:
```diff
@@ -0,0 +1,137 @@
+        super(HttpApi, self).__init__(connection)
+        self.connection = connection
+        self.sid = None
+        self._ignore_http_errors = False
```
```suggestion
```

This isn't actually doing anything anymore

Qalthos commented on lib/ansible/plugins/httpapi/checkpoint.py:
```diff
@@ -0,0 +1,137 @@
+class HttpApi(HttpApiBase):
+    def __init__(self, connection):
+        super(HttpApi, self).__init__(connection)
+        self.connection = connection
```
```suggestion
```

This is in the base class

Qalthos commented on lib/ansible/plugins/httpapi/checkpoint.py:
```diff
@@ -0,0 +1,137 @@
+        self.refresh_token = None
+        self.access_token = None
+
+    def _send_auth_request(self, path, data, **kwargs):
```
You haven't kept the overridden `handle_httperror`, so setting `_ignore_http_errors` doesn't do anything. I would probably remove this method.

Qalthos commented on lib/ansible/plugins/httpapi/checkpoint.py:
```diff
@@ -0,0 +1,137 @@
+            raise AnsibleConnectionFailure('Username and password are required for login')
+
+        try:
+            self.sid = response['sid']
```
Is `sid` usable as an auth token? if so,

```suggestion
            self._auth = {'X-chkp-sid': response['sid']}
```

will automatically attach the correct value to any request's headers (so you don't have to add it in `send_request`) and stops httpapi from also sending HTTP basic auth with every request

Qalthos commented on lib/ansible/plugins/httpapi/checkpoint.py:
```diff
@@ -0,0 +1,137 @@
+            return error['code'], error['message']
+
+    def handle_httperror(self, exc):
+        return None
```
I don't think this is right. It is a little ill-described in the docstring, but I _think_ you would rather return `False` than `None`, and if that is the case, you probably don't even need to override this method.

rcarrillocruz committed:
Use auth instead of sid property and return False on handle_httperror method

rcarrillocruz committed:
Fix version in docstring

Qalthos commented on lib/ansible/plugins/httpapi/checkpoint.py:
```diff
@@ -0,0 +1,132 @@
+class HttpApi(HttpApiBase):
+    def __init__(self, connection):
+        super(HttpApi, self).__init__(connection)
+        self._auth = None
```
```suggestion
```

Qalthos replied:
Actually, you don't need `__init__` at all anymore

Qalthos commented on lib/ansible/plugins/httpapi/checkpoint.py:
```diff
@@ -0,0 +1,132 @@
+            raise AnsibleConnectionFailure('Username and password are required for login')
+
+        try:
+            self._auth = response['sid']
```
```suggestion
            self.connection._auth = {'X-chkp-sid': response['sid']}
```
Right, sorry, I told you the wrong thing for this

Qalthos commented on lib/ansible/plugins/httpapi/checkpoint.py:
```diff
@@ -0,0 +1,132 @@
+
+    def send_request(self, path, body_params):
+        data = json.dumps(body_params) if body_params else '{}'
+        headers = {'Content-Type': 'application/json', 'X-chkp-sid': self._auth}
```
```suggestion
        headers = {'Content-Type': 'application/json'}
```
`httpapi` will automatically merge `connection._auth` with the headers you provide.

Qalthos commented on lib/ansible/plugins/httpapi/checkpoint.py:
```diff
@@ -0,0 +1,132 @@
+            error = json.loads(e.read())
+            return error['code'], error['message']
+
+    def handle_httperror(self, exc):
```
If you remove this method, `httpapi` will automatically handle 401 errors with a relogin

Qalthos commented on lib/ansible/plugins/httpapi/checkpoint.py:
```diff
@@ -0,0 +1,132 @@
+        display.vvvv('Web Services: %s %s' % ('POST', self.connection._url))
+
+    @staticmethod
+    def _get_response_value(response_data):
```
I would just make these loose functions rather than `@staticmethod`s

Qalthos commented on lib/ansible/plugins/httpapi/checkpoint.py:
```diff
@@ -0,0 +1,132 @@
+        try:
+            return json.loads(response_text) if response_text else {}
+        # JSONDecodeError only available on Python 3.5+
+        except getattr(json.decoder, 'JSONDecodeError', ValueError):
```
Ugh, `JSONDecodeError` is a `ValueError`
```suggestion
        except ValueError:
```

Qalthos reviewed (changes requested):
I might be getting a little pedantic here, but I'd like this to be a sterling example of a proper httpapi plugin when we're done.

Qalthos commented on lib/ansible/plugins/httpapi/checkpoint.py:
```diff
@@ -0,0 +1,132 @@
+        False
+
+    def _display_request(self):
+        display.vvvv('Web Services: %s %s' % ('POST', self.connection._url))
```
You can now do
```suggestion
         self.connection.queue_message('vvvv', 'Web Services: %s %s' % ('POST', self.connection._url))
```
and remove `display` from the plugin

rcarrillocruz committed:
Remove constructor

rcarrillocruz committed:
Remove Accept from base headers

rcarrillocruz committed:
Do not override http error handler and assign Checkpoint sid to connection _auth

There is scaffolding in the base class to autoappend the token, given
it is assigned to connection _send

rcarrillocruz committed:
Use new connection queue message method instead of display

rcarrillocruz committed:
Remove unused display

rcarrillocruz committed:
Catch ValueError, since it's a parent of JSONDecodeError

rcarrillocruz committed:
Make static methods that are not used outside the class regular methods

rcarrillocruz committed:
Add missing self to previously static methods

rcarrillocruz committed:
Fix logout

Was carrying copy pasta from ftd plugin

rcarrillocruz committed:
Remove send_auth_request

rcarrillocruz committed:
Use BASE_HEADERS constant

rcarrillocruz committed:
Simplify copyright header on httpapi plugin

rcarrillocruz committed:
Remove access rule module

rcarrillocruz committed:
Remove unused imports

rcarrillocruz committed:
Add unit test

rcarrillocruz committed:
Fix pep8

rcarrillocruz committed:
Add test

rcarrillocruz committed:
Add test

rcarrillocruz committed:
Fix pep8
